docs/docs/reference/other-new-features/explicit-nulls.md

```diff
@@ -0,0 +1,296 @@
+---
+layout: doc-page
+title: "Explicit Nulls"
+---
+
+This proposal describes a modification to the Scala type system that makes reference types
+(anything that extends `AnyRef`) _non-nullable_.
+
+This means the following code will no longer typecheck:
+```
+val x: String = null // error: found `Null`,  but required `String`
+```
+
+Instead, to mark a type as nullable we use a [type union](https://dotty.epfl.ch/docs/reference/new-types/union-types.html)
+```
+val x: String|Null = null // ok
+```
+
+Explicit nulls are enabled via a `-Yexplicit-nulls` flag, so they're an opt-in feature.
+
+Read on for details.
+
+## New Type Hierarchy
+
+When explicit nulls are enabled, the type hierarchy changes so that `Null` is subtype only of
+`Any`, as opposed to every reference type.
+
+This is the new type hierarchy:
+![](../../images/explicit-nulls/explicit-nulls-type-hierarchy.png "Type Hierarchy for Explicit Nulls")
+
+After erasure, `Null` remains a subtype of all reference types (as forced by the JVM).
+
+## Unsoundness
+
+The new type system is unsound with respect to `null`. This means there are still instances where an expressions has a non-nullable type like `String`, but its value is `null`.
+
+The unsoundness happens because uninitialized fields in a class start out as `null`:
+```scala
+class C {
+  val f: String = foo(f)
+  def foo(f2: String): String = if (f2 == null) "field is null" else f2
+}
+val c = new C()
+// c.f == "field is null"
+```
+
+Enforcing sound initialization is a non-goal of this proposal. However, once we have a type
+system where nullability is explicit, we can use a sound initialization scheme like the one
+proposed by @liufengyun and @biboudis in [https://github.com/lampepfl/dotty/pull/4543](https://github.com/lampepfl/dotty/pull/4543) to eliminate this particular source of unsoundness.
+
+## Equality
+
+Because of the unsoundness, we need to allow comparisons of the form `x == null` or `x != null`
```

Comment:
can only compare for equality nullable types
if want to compare non-null for some reason, can use cast (x: Any) != null

Reply:
also mention that can't use reference equality with null